suitesparse_bindings/sprs_suitesparse_umfpack/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+                let mut x = vec![0.0; ncol];
+
+                // Get C-compatible raw pointers to column pointers, indices, and values
+                let a = self.a();
+                let ap = a.indptr().to_proper().as_ptr();
+                let ai = a.indices().as_ptr();
+                let ax = a.data().as_ptr();
+
+                // Do the linear solve using pre-computed LU factors and permutation
+                unsafe {
+                    $solve(
+                        0,  // sys=0 for Ax=b problem type
+                        ap,
+                        ai,
+                        ax,
+                        x[..].as_mut_ptr(),
+                        b.as_ptr(),
+                        self.numeric.0,
+                        null() as *const f64,  // Default settings
+                        null_mut() as *mut c_void  // Ignore info
+                    )
+                };
+
+                x
+            }
+
+            /// Get shape info about LU components
+            /// * `lnz`: number of nonzero entries in `L`
+            /// * `unz`: number of nonzero entries in `U`
+            /// * `nrow`: number of rows in `L` and `U`
+            /// * `ncol`: number of columns in `L` and `U`
+            /// * `nz_udiag`: number of nonzeroes on the diagonal of `L` and `U`
+            pub fn get_lunz(&self) -> ($int, $int, $int, $int, $int) {
+                let mut lnz: $int = 0;  // Total number of nonzero entries in L
+                let mut unz: $int = 0;  // Total number of nonzero entries in U
+                let mut nrow: $int = 0;  // Number of rows in both L and U (should match input matrix)
+                let mut ncol: $int = 0;  // Number of cols in both L and U (should match input matrix)
+                let mut nz_udiag: $int = 0;  // Number of nonzero entries on the diagonal of each L and U
+
+                unsafe {
+                    $get_lunz(
+                        &mut lnz as *mut $int,
+                        &mut unz as *mut $int,
+                        &mut nrow as *mut $int,
+                        &mut ncol as *mut $int,
+                        &mut nz_udiag as *mut $int,
+                        self.numeric.0
+                    );
+                }
+
+                (lnz, unz, nrow, ncol, nz_udiag)
+            }
+
+            /// Get raw components of the numerical factorization of `A`
+            /// * `l`: `L` matrix in CSC format
+            /// * `u`: `U` matrix in CSR format
+            /// * `p`: row permutation
+            /// * `q`: column permutation
+            /// * `rs`: inverse row scaling (divide rows of LU by these to recover PAQ)
+            /// * `dx`: unknown; this quantity is not mentioned in underlying documentation but has distinct values, so we provide it here
+            pub fn get_numeric(&self) -> (CsMatI<f64, $int>, CsMatI<f64, $int>, PermOwnedI<$int>, PermOwnedI<$int>, Vec<f64>, Vec<f64>) {
```

Comment:
Should we return a struct from this function?

Reply:
I'm down with either option - this method allows you to mask out parts of the return which could reduce peak heap usage in cases where the compiler is able to identify that the values are never used... but for real applications you definitely need all of them except maybe dx, so it might as well be a struct

Reply:
Reworked it to return a struct. Definitely cleaner this way